In [104]:
import pandas as pd
import numpy as np

In [2]:
df = pd.read_csv('HW1-data.csv')

In [3]:
df.head()

User  Gender (1 =F, 0=M)  260: Star Wars: Episode IV - A New Hope (1977)  \
0   755                   0                                             1.0   
1  5277                   0                                             5.0   
2  1577                   1                                             NaN   
3  4388                   0                                             NaN   
4  1202                   1                                             4.0   

   1210: Star Wars: Episode VI - Return of the Jedi (1983)  \
0                                                5.0         
1                                                3.0         
2                                                NaN         
3                                                3.0         
4                                                3.0         

   356: Forrest Gump (1994)  318: Shawshank Redemption, The (1994)  \
0                       2.0                                    NaN   
1                       NaN                                    2.0   
2                       NaN                                    5.0   
3                       NaN                                    NaN   
4                       4.0                                    1.0   

   593: Silence of the Lambs, The (1991)  3578: Gladiator (2000)  \
0                                    4.0                     4.0   
1                                    4.0                     2.0   
2                                    2.0                     NaN   
3                                    NaN                     1.0   
4                                    4.0                     1.0   

   1: Toy Story (1995)  2028: Saving Private Ryan (1998)  ...  \
0                  2.0                               2.0  ...   
1                  1.0                               NaN  ...   
2                  4.0                               NaN  ...   
3                  2.0                               3.0  ...   
4                  NaN                               4.0  ...   

   2396: Shakespeare in Love (1998)  2916: Total Recall (1990)  \
0                               2.0                        NaN   
1                               3.0                        2.0   
2                               NaN                        1.0   
3                               NaN                        4.0   
4                               5.0                        1.0   

   780: Independence Day (ID4) (1996)  541: Blade Runner (1982)  \
0                                 5.0                       2.0   
1                                 2.0                       NaN   
2                                 4.0                       4.0   
3                                 1.0                       3.0   
4                                 NaN                       4.0   

   1265: Groundhog Day (1993)  2571: Matrix, The (1999)  \
0                         NaN                       4.0   
1                         2.0                       NaN   
2                         1.0                       1.0   
3                         5.0                       NaN   
4                         NaN                       3.0   

   527: Schindler's List (1993)  2762: Sixth Sense, The (1999)  \
0                           2.0                            5.0   
1                           5.0                            1.0   
2                           2.0                            3.0   
3                           5.0                            1.0   
4                           5.0                            5.0   

   1198: Raiders of the Lost Ark (1981)  34: Babe (1995)  
0                                   NaN              NaN  
1                                   3.0              NaN  
2                                   1.0              3.0  
3                                   1.0              2.0  
4                                   NaN              NaN  

[5 rows x 22 columns]

1. Mean Rating: Calculate the mean rating for each movie, order with the highest rating listed first, and submit the top three (along with the mean scores for the top two).

In [17]:
movies = df.columns[2:]
df[movies].mean().sort_values(ascending=False).head(3)

318: Shawshank Redemption, The (1994)             3.600000
260: Star Wars: Episode IV - A New Hope (1977)    3.266667
541: Blade Runner (1982)                          3.222222
dtype: float64

2. Rating Count (popularity): Count the number of ratings for each movie, order with the most number of ratings first, and submit the top three (along with the counts for the top two).

In [18]:
df[movies].notnull().sum().sort_values(ascending=False).head(3)

1: Toy Story (1995)                               17
593: Silence of the Lambs, The (1991)             16
260: Star Wars: Episode IV - A New Hope (1977)    15
dtype: int64

3. % of ratings 4+ (liking): Calculate the percentage of ratings for each movie that are 4 or higher. Order with the highest percentage first, and submit the top three (along with the percentage for the top two).  Notice that the three different measures of "best" reflect different priorities and give different results; this should help you see why you need to be thoughtful about what metrics you use. 

In [86]:
((df[movies] >= 4).sum() / df[movies].notnull().sum()).sort_values(ascending=False).head(3)

318: Shawshank Redemption, The (1994)             0.700000
260: Star Wars: Episode IV - A New Hope (1977)    0.533333
3578: Gladiator (2000)                            0.500000
dtype: float64

4. Top movies for someone who has seen Toy Story: Calculate movies that most often occur with Movie #1:  Toy Story, using the (x+y)/x method described in class. In other words, for each movie, calculate the percentage of Toy Story raters who also rated that movie. Order with the highest percentage first, and submit the top 3 (along with the correlations for the top two).  Note, you will have ties -- to break the ties, use the lowest-numbered movie as the higher-ranked one.  In other words, if Movies 541 and 318 are tied, then 318 gets the higher rank.

In [49]:
ranked_toy_story = df['1: Toy Story (1995)'].notnull()

(df.loc[ranked_toy_story, movies].notnull().sum().sort_values(ascending=False) / ranked_toy_story.sum()).iloc[1:5]

260: Star Wars: Episode IV - A New Hope (1977)             0.823529
593: Silence of the Lambs, The (1991)                      0.764706
1210: Star Wars: Episode VI - Return of the Jedi (1983)    0.764706
780: Independence Day (ID4) (1996)                         0.764706
dtype: float64

5. Correlation with Toy Story:  Calculate the correlation between the vectors of ratings for Toy Story and each other movie.  You can use the built-in CORREL() function.  Order by the highest correlation (positive only) and submit the top 3 along with the correlation values for the top 2.  Notice the differences between co-occurrence and correlation; these metrics are showing different types of relationships.


In [50]:
df[movies].corr().loc['1: Toy Story (1995)'].sort_values(ascending=False).iloc[1:4]

318: Shawshank Redemption, The (1994)    0.888523
34: Babe (1995)                          0.811107
296: Pulp Fiction (1994)                 0.709842
Name: 1: Toy Story (1995), dtype: float64

6. Mean rating difference by gender:  First, recompute the mean rating for each movie separately for males and for females.  And calculate the overall mean rating (across all ratings) for males and females. Submit the two movies that have the greatest differences (one where men are most above women, and one where women are most above men) along with the differences in average.  Also submit the difference in overall rating averages (female average - male average).


In [53]:
df_means = df.drop(columns='User').groupby('Gender (1 =F, 0=M)').mean()

# overall mean rating 
df_means.mean(axis=1)

Gender (1 =F, 0=M)
0    2.884286
1    2.999286
dtype: float64

In [66]:
# movies with the most difference
(df_means.loc[1] - df_means.loc[0]).sort_values().iloc[[0, -1]]

1198: Raiders of the Lost Ark (1981)   -1.666667
2396: Shakespeare in Love (1998)        2.107143
dtype: float64

In [111]:
female = df['Gender (1 =F, 0=M)'] == 1 

rates_f = df.loc[female, movies].values.flatten()
rates_f = rates_f[~np.isnan(rates_f)]


rates_m = df.loc[~female, movies].values.flatten()
rates_m = rates_m[~np.isnan(rates_m)]


rates_f.mean() - rates_m.mean()

np.float64(0.0418566100290092)

7. Next, compute the % of ratings 4+ separately for males and females.  You'll be asked to submit two movies as above (largest difference in each direction). And again you'll indicate whether men or women are more likely to rate movies 4 stars or above.


In [124]:
df_f_4plus = (df.loc[female, movies] >= 4).sum() / df.loc[female, movies].notnull().sum()
df_m_4plus = (df.loc[~female, movies] >= 4).sum() / df.loc[~female, movies].notnull().sum()

In [135]:
# movies with the biggest gap 
(df_f_4plus - df_m_4plus).sort_values(ascending=False).iloc[[0, -1]]

2396: Shakespeare in Love (1998)        0.75
1198: Raiders of the Lost Ark (1981)   -0.50
dtype: float64